### PR TITLE
storage: use race-free `AddNames` instead of `SetNames`

### PR DESCRIPTION
**Pull request by flouthoc**
Commits from parallel builds using `SetNames` removes `names` from
storage for other builds.

Use race-free atomic `AddNames` to prevent breaking of parallel builds.

Uses `c/storage` from PR: https://github.com/containers/storage/pull/1153

Fixes issue: 

**Builds fail with image not known.**
### Reproducer
```console
#!/bin/bash
x=1
sudo ./podman rmi -af
rm -f log*.*
sudo ./podman build -t first . &> logfirst.log
while [ $x -le 30 ]
do
  echo "$x times"
  sudo ./podman build --log-level debug -t $x . &> log$x.log &
  x=$(( $x + 1 ))
#  sleep 1
done
```

### Dockerfile
```console
FROM quay.io/jitesoft/alpine
```

PS: I am not sure how to ensure parallel builds in CI and produce a race so not sure how to test this in CI.

### CODE REVIEW

**Comment by flouthoc** [comment]
I have to `point` to actual `c/storage` version right now its my fork.

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1207]: The `oldNames` variable is now conceptually unnecessary and should be eliminated (_maybe_ apart from the error message about cleanup).

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1208]: … Then the `names` variable will have a trivial value and can be eliminated/inlined.

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1213]: All of these texts should only list the newly added names now; they shouldn’t list a set of names that might never actually exist in storage.

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1207]: Also update the `SetNames()` comment above.

Comment (flouthoc) [reply storage/storage_image.go]: Actually i wanted to be change to be minimum like just `s/SetNames/AddNames` but you are right conceptually we don't need `oldNames` anymore.

File: storage/storage_image.go
Comment (vrothberg) [inline storage/storage_image.go:1208]: No need for a log when returning the same error.

File: storage/storage_image.go
Comment (vrothberg) [inline storage/storage_image.go:1204]: This check can be removed; we already know that `name != nil`.

File: storage/storage_image.go
Comment (vrothberg) [inline storage/storage_image.go:1211]:
```suggestion
		logrus.Debugf("add name %q to image %s",  name, img.ID)
```

File: storage/storage_image.go
Comment (vrothberg) [inline storage/storage_image.go:1205]: With the two lines above deleted.
```suggestion
		if err := s.imageRef.transport.store.AddNames(img.ID, string[]{name.String()}); err != nil {
```

**Review by vrothberg — COMMENTED** [review]
LGTM

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1204]:
```suggestion
			return errors.Wrapf(err, "adding names %v to image %q", name, img.ID)
```

File: storage/storage_image.go
Comment (mtrmac) [inline storage/storage_image.go:1206]:
```suggestion
		logrus.Debugf("added name %q to image %s", name, img.ID)
```
Also, `img.ID` originally used `%q`, and I don’t currently see a reason to change that. (I don’t have a strong opinion on use of `%q` for `name`.)

**Comment by flouthoc** [comment]
Wait for https://github.com/containers/storage/pull/1153 and tests to pass here: https://github.com/containers/podman/pull/13339

File: storage/storage_image.go
Comment (mtrmac) [reply storage/storage_image.go]: FWIW the way I think about this, there will never be more time or more motivation to clean up and simplify the code.

Keeping the code as simple and as easy to maintain as possible is an everyday job.

**Review by mtrmac — APPROVED** [review]
LGTM — just so that it isn‘t forgotten, the `replace` in `go.mod` should be replaced.

After that is done, and tests pass, feel free to merge without another review.

**Comment by mtrmac** [comment]
Also: Thanks!

**Review by vrothberg — APPROVED** [review]
LGTM

File: go.mod
Comment (mtrmac) [inline go.mod:46]: The dependency above needs to be updated, without using `replace`; `replace` is ignored when importing c/image into other users.

Comment (vrothberg) [reply go.mod:46]: thanks for catching that, @mtrmac !

**Comment by flouthoc** [comment]
@mtrmac @vrothberg done.

**Comment by mtrmac** [comment]
LGTM. Can you rebase it over the release, please, so that it’s trivially clear which commits part of 5.20.0 and which commits come after? Feel free to merge afterwards.

**Review by vrothberg — APPROVED** [review]
LGTM

**Comment by flouthoc** [comment]
@containers/image-maintainers @mtrmac @vrothberg  I want to backport this commit to `v5.16.0` but I was unable to find any maintenance branch for that. Is there any branch where i could raise a PR to get this backported. 

Thanks

**Comment by vrothberg** [comment]
@flouthoc, I just created one: https://github.com/containers/image/tree/release-5.16

When backporting, please make sure to bump the version accordingly in `version/version.go`.

**Review by mtrmac — COMMENTED** [review]
ACK to the code (but blocked on a c/storage release).

WRT the test failures, is it possible to fix that by updating `SKOPEO_BRANCH` on this branch to an appropriate consumer of c/image 5.16?

**Pull request merged**
